src/lib/TopLevel.hs

```diff
@@ -193,8 +195,17 @@ evalBackend expr = do
       checkPass ImpPass impFunction
       logPass Flops $ impFunctionFlops impFunction
       let llvmFunc = impToLLVM impFunction
-      outVars <- liftIO $ execLLVM logger engine llvmFunc inVars
-      return $ reStructureArrays (getType expr) $ map Var outVars
+      let exprType = getType expr
+      lift $ modifyMVar engine $ \env -> do
+        let inPtrs = fmap (env !) inVars
+        outPtrs <- callLLVM logger llvmFunc inPtrs
+        let outTys = flattenType exprType
+        -- TODO: Assert length of outPtrs equal to outTys
+        let (outNames, env') = nameItems (Name ArrayName "arr" 0) env outPtrs
+        let outVars = zipWith (\v (_, b) -> v :> ArrayTy b) outNames outTys
+        -- XXX: This assumes that the implementation of flattenType and makeDest in Imp yield the same order!
```

Comment:
TODO: It should be the responsibility of the backend to return the atom with the appropriate type. Right now we assume that the flattening implementations from `Imp` and `JAX` can always match exactly with the unflattening implemented in `Serialize`